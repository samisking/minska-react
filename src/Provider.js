import React from 'react';

class Provider extends React.PureComponent {
  static propTypes = {
    children: React.PropTypes.node.isRequired,
    store: React.PropTypes.object.isRequired
  };

  static childContextTypes = {
    store: React.PropTypes.object
  };

  state = {
    store: this.props.store
  };

  getChildContext() {
    return {
      store: this.state.store
    };
  }

  render() {
    return (
      this.props.children
    );
  }
}

export default Provider;
