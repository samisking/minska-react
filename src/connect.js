import React from 'react';

const defaultMapStateToProps = state => (state);

const Connect = (
  mapStateToProps = defaultMapStateToProps
) => WrappedComponent => class extends React.PureComponent {
  static contextTypes = {
    store: React.PropTypes.object
  }

  constructor(props, context) {
    super(props, context);

    this.storeKey = `connect:_${Math.random().toString(36).substr(2, 9)}`;
    this.onUpdate = this.onUpdate.bind(this);
    this.send = this.send.bind(this);

    const store = context.store;

    if (!store) {
      throw new Error('Store could not be found in connected component. Make sure to wrap your ' +
                      'root component in <Provider> with a store prop.');
    }

    this.state = {
      ...context.store.state
    };
  }

  componentWillMount() {
    this.context.store.subscribe('onChange', this.storeKey, this.onUpdate);
  }

  componentWillUnmount() {
    this.context.store.unsubscribe(this.storeKey);
  }

  onUpdate(state) {
    this.setState(state);
  }

  send(...args) {
    this.context.store.send(...args);
  }

  render() {
    const stateToProps = mapStateToProps(this.state, this.props);

    return (
      <WrappedComponent
        {...this.props}
        {...stateToProps}
        send={this.send}
      />
    );
  }
};

export default Connect;
