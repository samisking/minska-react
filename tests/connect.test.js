import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Store from 'minska';
import MyApp from './mockApp';
import { Provider, connect } from '../src';

let store;

beforeEach(() => {
  store = new Store({
    state: {
      foo: 'bar'
    },
    reducers: {
      updateFoo: (state, foo) => ({ ...state, foo })
    }
  });
});

// The actual tests
describe('connect', () => {
  it('renders correctly', () => {
    const ConnectedApp = connect()(MyApp);
    const component = renderer.create(
      <Provider store={store}>
        <ConnectedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('passes correct props', () => {
    const ConnectedApp = connect()(MyApp);
    const component = mount(
      <Provider store={store}>
        <ConnectedApp />
      </Provider>
    );

    expect(component.find(MyApp).props()).toMatchSnapshot();
  });
});
