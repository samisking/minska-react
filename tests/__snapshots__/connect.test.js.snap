// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect passes correct props 1`] = `
Object {
  "foo": "baz",
  "send": [Function],
}
`;

exports[`connect renders correctly 1`] = `
<div>
  Foo is 
  baz
</div>
`;
